batch/batch/front_end/query/query_v2.py

```diff
@@ -9,20 +9,157 @@
+                    ):
+                        min_start_gt_query = st_query
+                elif left == 'end_time':
+                    et_query = BatchEndTimeQuery.parse(op, right)
+                    queries.append(et_query)
+                    if (type(et_query.operator) in [LessThanOperator, LessThanEqualOperator]) and (
+                        max_end_lt_query is None or max_end_lt_query.time_msecs <= et_query.time_msecs
+                    ):
+                        max_end_lt_query = et_query
+                elif left == 'duration':
+                    queries.append(BatchDurationQuery.parse(op, right))
+                elif left == 'cost':
+                    queries.append(BatchCostQuery.parse(op, right))
+                else:
+                    queries.append(BatchKeywordQuery.parse(op, left, right))
+            else:
+                raise QueryError(f'could not parse term "{_term}"')
+
+    # this is to make time interval queries fast by using the bounds on both indices
+    if min_start_gt_query and max_end_lt_query and min_start_gt_query.time_msecs <= max_end_lt_query.time_msecs:
+        queries.append(BatchStartTimeQuery(max_end_lt_query.operator, max_end_lt_query.time_msecs))
+        queries.append(BatchEndTimeQuery(min_start_gt_query.operator, min_start_gt_query.time_msecs))
+
+    # batch has already been validated
+    where_conditions = [
+        '(billing_project_users.`user` = %s)',
+        'NOT deleted',
+    ]
+    where_args: List[Any] = [user]
+
+    if last_batch_id is not None:
+        where_conditions.append('(batches.id < %s)')
+        where_args.append(last_batch_id)
+
+    for query in queries:
+        cond, args = query.query()
+        where_conditions.append(f'({cond})')
+        where_args += args
+
+    sql = f'''
+SELECT batches.*,
+  batches_cancelled.id IS NOT NULL AS cancelled,
+  batches_n_jobs_in_complete_states.n_completed,
+  batches_n_jobs_in_complete_states.n_succeeded,
+  batches_n_jobs_in_complete_states.n_failed,
+  batches_n_jobs_in_complete_states.n_cancelled,
+  cost_t.cost
+FROM batches
+LEFT JOIN billing_projects ON batches.billing_project = billing_projects.name
+LEFT JOIN batches_n_jobs_in_complete_states ON batches.id = batches_n_jobs_in_complete_states.id
+LEFT JOIN batches_cancelled ON batches.id = batches_cancelled.id
+STRAIGHT_JOIN billing_project_users ON batches.billing_project = billing_project_users.billing_project
+LEFT JOIN LATERAL (
+  SELECT COALESCE(SUM(`usage` * rate), 0) AS cost
+  FROM (
+    SELECT batch_id, resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
+    FROM aggregated_batch_resources_v2
+    WHERE batches.id = aggregated_batch_resources_v2.batch_id
+    GROUP BY batch_id, resource_id
+  ) AS usage_t
+  LEFT JOIN resources ON usage_t.resource_id = resources.resource_id
+  GROUP BY batch_id
+) AS cost_t ON TRUE
+WHERE {' AND '.join(where_conditions)}
+ORDER BY id DESC
+LIMIT 51;
+'''
+
+    sql_args = where_args
```

Comment:
nit: this line doesn't really do anything, I think we should just `return (sql, where_args)`